include/cuco/detail/common_kernels.cuh

```diff
@@ -76,7 +76,7 @@ __global__ void insert_if_n(InputIterator first,
 
   while (idx < n) {
     if (pred(*(stencil + idx))) {
-      typename Ref::value_type const insert_element{*(first + idx)};
+      auto const insert_element{*(first + idx)};
```

Comment:
I think we started with `auto` and ran into issues with thrust raw reference cast. Could use something like:
```cpp
// raw reference cast might be needed as well
typename std::iterator_traits<InputIterator>::value_type const insert_element{*(first + idx)};
```

Reply:
Yeah, I might need to solve the `raw_reference_cast` problem first in order to get this working. The question is where to place these casts and do they introduce any overheads?

Reply:
> The question is where to place these casts 

They should be at the kernel level right before passing the key/value to the device ref APIs.

> and do they introduce any overheads?

Can be easily checked with godbolt and should be no more expensive than a copy.

Reply:
> They should be at the kernel level right before passing the key/value to the device ref APIs.

I was thinking of putting this cast at the top of each device operator, since otherwise the device API cannot benefit from this feature.

> Can be easily checked with godbolt and should be no more expensive than a copy.

Will do. In the worst case the result of the cast will occupy an additional register, but I'm confident the compiler is capable of optimizing this out.

Reply:
> I was thinking of putting this cast at the top of each device operator, since otherwise the device API cannot benefit from this feature.

I see, it shouldn't affect much though in terms of runtime performance.

Then if users choose to use device APIs, wouldn't it be the users' responsibility to dereference the iterator and pass a proper type to the device ref APIs?